lesson03.find-a-pair/index.html

```diff
@@ -1,22 +1,358 @@
+                shuffle(this.cards);
+
+                // assign cards to html
+                this.cells = [...document.querySelectorAll('.cell')];
+                for (let i = 0; i < this.numCards; i++) {
+                    this.cells[i].classList.add(this.cards[i].type);
+                }
+
+                // add click event listener
+                this.cells.forEach(cell => {
+                    cell.onclick = () => game.openCard(cell);
+                });
+
+
+                // show image for 4 seconds to remember
+                window.setTimeout(function () {
+                    game.cells.forEach(cell => {
+                        cell.className = 'cell unmatched';
+                    });
+                }, 4000);
+            }
+
+            openCard(selectedCell) {
+                // show image
+                const index = parseInt(selectedCell.getAttribute('index'));
+                const selectedCard = this.cards[index]
+                selectedCell.classList.add(selectedCard.type);
+
+                if (this.openedCardIndex < 0) {
+                    // first card to open, keep showing the image
+                    this.openedCardIndex = index;
+                }
+                else {
+                    const openedCell = this.cells[this.openedCardIndex];
+
+                    // try to match
+                    const openedCard = this.cards[this.openedCardIndex];
+                    const isMatched = openedCard.isMatch(selectedCard);
+                    if (!isMatched) {
+                        // close cards
+                        window.setTimeout(function () {
+                            game.closeUnmatched(openedCell, selectedCell);
+                        }, 1000);
+                    }
+                    else {
+                        // remove from board
+                        window.setTimeout(function () {
+                            game.removeMatched(openedCell, selectedCell);
+                        }, 1000);
+                    }
+                    this.openedCardIndex = -1;
+                }
+            }
+
+            closeUnmatched(cell1, cell2) {
+                cell1.className = 'cell unmatched';
+                cell2.className = 'cell unmatched';
+            }
+
+            removeMatched(cell1, cell2) {
+                cell1.className = 'cell matched';
+                cell2.className = 'cell matched';
+                cell1.onclick = null;
+                cell2.onclick = null;
+            }
+        }
+
+        // Base card
+        class Card {
+            constructor(type) {
+                this.type = type;
+            }
+
+            isMatch(card) {
+                return this.type === card.type;
+            }
+        }
+
+        class Dog extends Card {
+            constructor() {
+                super("dog");
+            }
+        }
+
+        class Cat extends Card {
+            constructor() {
```

Comment:
No need for constructor here - read about default constructors and inheritance in javascript